internal/jimm/model_cleanup.go

```diff
@@ -0,0 +1,59 @@
+// Copyright 2024 Canonical.
+
+package jimm
+
+import (
+	"context"
+	"fmt"
+
+	jujuparams "github.com/juju/juju/rpc/params"
+	"github.com/juju/juju/state"
+	"github.com/juju/zaputil/zapctx"
+	"go.uber.org/zap"
+
+	"github.com/canonical/jimm/v3/internal/dbmodel"
+	"github.com/canonical/jimm/v3/internal/errors"
+)
+
+// CleanupModelsDying loops over dying models, contacting the respective controller.
+// And deleting the model from our database if the error is `NotFound` which means the model was successfully deleted.
+func (j *JIMM) CleanupModelsDying(ctx context.Context) error {
+	const op = errors.Op("jimm.WatchModelsDying")
+
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	err := j.DB().ForEachModel(ctx, func(m *dbmodel.Model) error {
+		if m.Life == state.Dying.String() {
+			// if the model is dying and not found by querying the controller we can assume it is dead.
+			// And safely delete the reference from our db.
+			api, err := j.dialModel(ctx, &m.Controller, m.ResourceTag())
```

Comment:
What about following the same pattern as the previous code. Something like:
1. Start a routine for each controller.
2. Open API connection to controller.
3. GetDyingModels(ThisController)
4. Query controller for model status.

If the routine returns with an error, the caller can reschedule after e.g. 30s. It just saves us for making a new connection for each model. Granted this may be a premature optimisation but it feels cleaner and aligns with how the watcher worked before anyway.